nextest-runner/src/test_output.rs

```diff
@@ -0,0 +1,322 @@
+            .sum();
+
+        self.chunks
+            .iter()
+            .fold(bytes::BytesMut::with_capacity(count), |mut acc, oc| {
+                if oc.stdout != stdout {
+                    return acc;
+                }
+
+                acc.extend_from_slice(&self.buf[oc.range.clone()]);
+                acc
+            })
+            .freeze()
+    }
+}
+
+/// Captures the stdout and/or stderr streams into a buffer, indexed on each
+/// chunk of output including the timestamp and which stream it came from
+pub struct TestOutputAccumulator {
+    buf: BytesMut,
+    chunks: Vec<OutputChunk>,
+    start: Instant,
+}
+
+impl TestOutputAccumulator {
+    /// Creates a new test accumulator to capture output from a child process
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            buf: BytesMut::with_capacity(4 * 1024),
+            chunks: Vec::new(),
+            start: Instant::now(),
+        }
+    }
+
+    /// Similar to [`bytes::BytesMut::freeze`], this is called when output
+    /// capturing is complete to create a [`TestOutput`] of the complete
+    /// captured output
+    pub fn freeze(self) -> TestOutput {
+        TestOutput {
+            buf: self.buf.freeze(),
+            chunks: self.chunks,
+            start: self.start,
+        }
+    }
+
+    /// Gets a writer the can be used to write to the accumulator as if a child
+    /// process was writing to stdout
+    #[inline]
+    pub fn stdout(&mut self) -> TestOutputWriter<'_> {
+        TestOutputWriter {
+            acc: self,
+            stdout: true,
+        }
+    }
+
+    /// Gets a writer the can be used to write to the accumulator as if a child
+    /// process was writing to stderr
+    #[inline]
+    pub fn stderr(&mut self) -> TestOutputWriter<'_> {
+        TestOutputWriter {
+            acc: self,
+            stdout: false,
+        }
+    }
+}
+
+/// Provides [`std::io::Write`] and [`std::fmt::Write`] implementations for a
+/// single stream of a [`TestOutputAccumulator`]
+pub struct TestOutputWriter<'acc> {
+    acc: &'acc mut TestOutputAccumulator,
+    stdout: bool,
+}
+
+impl<'acc> std::io::Write for TestOutputWriter<'acc> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let start = self.acc.buf.len();
+        self.acc.buf.extend_from_slice(buf);
+        self.acc.chunks.push(OutputChunk {
+            range: start..self.acc.buf.len(),
+            timestamp: Instant::now(),
+            stdout: self.stdout,
+        });
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+impl<'acc> std::fmt::Write for TestOutputWriter<'acc> {
+    fn write_str(&mut self, s: &str) -> std::fmt::Result {
+        match self.write(s.as_bytes()) {
+            Ok(_) => Ok(()),
+            Err(_) => Err(std::fmt::Error),
+        }
+    }
+}
+
+pin_project! {
+    #[project = StreamProj]
+    enum Stream {
```

Comment:
I think you could use [`tokio_util::either::Either`](https://docs.rs/tokio-util/latest/tokio_util/either/enum.Either.html) here, it is probably what you want and it implements `AsyncRead` and `AsyncBufRead`.